src/main/java/com/openshift/internal/restclient/capability/resources/OpenShiftBinaryPodLogRetrieval.java

```diff
@@ -39,42 +47,114 @@ public boolean isSupported() {
 	public String getName() {
 		return OpenShiftBinaryPodLogRetrieval.class.getSimpleName();
 	}
-
+	
 	@Override
 	public InputStream getLogs(boolean follow) {
-		this.follow = follow;
-		start();
-		SequenceInputStream is = new SequenceInputStream(getProcess().getInputStream(), getProcess().getErrorStream());
-		return is;
+		return getLogs(follow, null);
 	}
 
 	@Override
-	protected void cleanup() {
-		follow = false;
-		if(getProcess() != null) {
-			IOUtils.closeQuietly(getProcess().getInputStream());
-			IOUtils.closeQuietly(getProcess().getErrorStream());
+	public InputStream getLogs(boolean follow, String container) {
+		container = StringUtils.defaultIfBlank(container, "");
+		synchronized (cache) {
+			if(cache.containsKey(container)) {
+				return cache.get(container).getLogs();
+			}
+			PodLogs logs = null;
+			try {
+				logs = new PodLogs(client, follow, container, cache::remove);
+				return logs.getLogs();
+			}catch(Exception e) {
+				throw e;
+			}finally {
+				if(logs != null) {
+					cache.put(container, logs);
+				}
+			}
 		}
 	}
-
+	
 	@Override
-	protected boolean validate() {
-		return true;
+	public void stop() {
+		new ArrayList<>(cache.keySet()).forEach(c->stop(c));
 	}
 
 	@Override
-	protected String buildArgs() {
-		StringBuilder args = new StringBuilder();
-		args.append("logs ");
-		addSkipTlsVerify(args);
-		addServer(args)
-		.append(" ").append(pod.getName()).append(" ")
+	public synchronized void stop(String container) {
+		if(!cache.containsKey(container)) return;
+		try {
+			PodLogs logs = cache.remove(container);
+			logs.stop();
+		}catch(Exception e) {
+			LOG.warn("Unable to stop pod logs",e);
+		}
+	}
+
+
+	private class PodLogs extends AbstractOpenShiftBinaryCapability{
+		
+		private String container;
+		private boolean follow;
+		private Consumer<String> cacheFlush;
+		private SequenceInputStream is;
+
+		PodLogs(IClient client, boolean follow, String container, Consumer<String> cacheFlush){
+			super(client);
+			this.follow = follow;
+			this.container = container;
+			this.cacheFlush = cacheFlush;
+		}
+
+		public synchronized InputStream getLogs() {
+			if(is == null) {
+				start();
+				is = new SequenceInputStream(getProcess().getInputStream(), getProcess().getErrorStream());
+			}
+			return is;
+		}
+
+		@Override
+		public boolean isSupported() {
+			return true;
+		}
+
+		@Override
+		public String getName() {
+			return "";
+		}
+
+		@Override
+		protected void cleanup() {
+			follow = false;
+			if(getProcess() != null) {
+				IOUtils.closeQuietly(getProcess().getInputStream());
+				IOUtils.closeQuietly(getProcess().getErrorStream());
+			}
+			cacheFlush.accept(this.container);
```

Comment:
It seems that this is not a synchronized call to cache.remove. Class PodLogs is inner in OpenShiftBinaryPodLogRetrieval so that it sees cache and can call 
synchronized(cache) { cache.remove(container); }